# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек
Выполнил студент гр. 0304 Жиглов Дмитрий, вариант 34.

## Цель работы
Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения
**Бинарный метод** - метод быстрого возведения числа в натуральную степень $n$.  
Алгоритм:  
1) Записать число $n$ в бинарном виде  
2) Отбросить старший бит  
3) Произвести замену: если бит равен единице, то заменить его на SX иначе заменить на S  
4) Выполнить вычисление, где S - возведение в квадрат, а X - умножение на $x$   
**Метод множителей** - метод быстрого возведения числа в натуральную степень n.  
Алгортим:  
1) Представляем $n = p*q$, где $p$ - наименьший простой множитель $n$, $q > 1$. Таким образом $x^n$ можно найти, вычислив $x^p$ и возведя эту величину в степень $q$.  
2) Если $n$ - простое, то можно сначала вычислить $x^{n-1}$ и умножить его на x  
3) При $n = 1$ получим $x^n$ безо всяких вычислений.  
**Аддитивной цепочкой** для натурального числа $n$ называется последовательность натуральных чисел $a_0 = 1,a_1,a_2,...,a_m = n$, где каждый элемент последовательности равен сумме каких-либо двух предыдущих:
\\[ a_i = a_j + a_k, \; k \leq j \leq i, \; i = 1,2,...,r \\]  
**Алгоритм Брауэра** - метод нахождения аддитивной цепочки для натурального числа $n$.  
Алгоритм:  
Для натурального числа $n$ при заданном начальном числе $k$ можно построить цепочку Брауэра с помощью рекурсивной формулы:
$\begin{equation*}
 B_k(n) = 
     \begin{cases}
       1,2,3,...,2^k - 1 &\text{if $n < 2^k$}\\
           B_k(q),2*q,4*q,8*q,...,2^k*q,n &\text{if $n >= 2^k,  q = [\frac{n}{2^k}]$}
     \end{cases}
    \end{equation*}$  
1) Задаётся фиксированный параметр $k$ для рассматриваемого числа $n$. Вычисляются "вспомогательные числа":
$d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \; mod \; d => n = q_1*d + r_1$  
$q_2 = [\frac{q1}{d}],r_2 = q_1 \; mod \; d => q_1 = q_2*d + r_2$  
2) Процедура продолжается до тех пор, пока не появится такое $q_s <  d => q_{s-1} = q_s*d + r_s$  
3) Таким образом $n$ имеет вид:
    \\[n = d^s*q_s + d^{s-1}*r_s + d^{s-2}*r_{s-1} + ... + \; r_1 \\]  
    **Алгоритм Яо** - метод нахождения аддитивной цепочки для натурального числа $n$.
    Алгоритм:
    Задаём некоторое целое $k >= 2$ и число $n$ раскладывается в $2^k$-й системе счисления:
    \\[ n = \sum\limits_{i=0}^j a_i*2^{i*k} \; a_j \neq 0 \\]
    Введём функцию d: 
    \\[d(z) = \sum_{i:a_i = z} 2^{i*k}\\]  
    1) Базовая последовательность:
    \\[1,2,4,8,...,2^{\lambda(n)}, \; где \; \lambda(n) - уменьшенная \; на \; единицу \; длина \; бинарной \; записи \; числа \; n \\]  
    2) Вычисление $d(z)$ для всех $z \in \{1,2,3,...,2^k-1\}, \; d(z) \neq 0$  
    3) Вычисление $z*d(z)$ для всех $z$  
    4) В конечном итоге, $n$ представляет собой разложение вида:
    \\[ n = \sum\limits_{z = 1}^{2^{k-1}} z*d(z) \\]  
    **Звёздной цепочкой** называется аддитивная цепочка включающая в себя только звёздные шаги.  
    Пара $(j,k), 0 \leq k \leq j < i \;$ называется  **шагом**  $i $. Тогда при $j = i-1$ пара называется **звёздным шагом**.

**Алгоритм дробления вектора индексов** - метод нахождения минимальной звёздной цепочки для натурального числа $n$.  
Алгоритм:  
1) Задаётся начальный вектор $r = \{1,2,3,...,m \}$ по которому строится начальная цепочка $a = \{a_1 = 1,a_2,a_3,...,a_{m+1}\}$  
2) Если $a_{m+1} = n$, то алгоритм завершается. В противном случае вектор r делится на две части: изменяемую и неизменяемую.  
3) Находим $a_{min}$ и $a_{max}$. Если $n \in [a_{min},a_{max}]$, то изменяемая часть вектора уменьшается на единицу по самому старшей позиции.  
4) Если цепочка так и не была найдена по всем возможным переборам изменяемой части вектора вплоть до $p = {1,1,...,1}$, то изменяемая часть принимает первоначальное значение, а неизменяемая уменьшается на единицу и процесс повторяется снова.  
5) Если были перебраны все варианты обоих частей вектора, то вектор $r$ увеличивается на единицу и принимает значение $r = {1,2,...,m+1}$.  
6) Алгоритм продолжается, пока не будет найдена цепочка.

## Постановка задачи
Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы
1) Были выполнены вычисления для $x^n$ бинарным методом и методом множителей для значений $n = 60, 139$:  
**n = 60**  
Двоичный метод "SX":
    \\[ 60 = 111100_2 \\]
    \\[ 11100_2 = SXSXSXSS \\]
    \\[ SXSXSXSS = x^2,x^3,x^6,x^7,x^{14},x^{15},x^{30},x^{60} \\]  
    Метод множителей:
    \\[ 60 = p*q = 2*30\\]
    \\[ y = x^2, \; y^{30} = (y^5)^6 = ((y^2*y^2*y)^2)^2*(y^2*y^2*y)^2 \\]  
    **n = 139**  
    Двоичный метод "SX":
    \\[ 139 = 10001011_2 \\]
    \\[ 0001011_2 = SSSSXSSXSX \\]
    \\[ SSSSXSSXSX = x^2,x^4,x^8,x^{16},x^{17},x^{34},x^{68},x^{69},x^{138},x^{139} \\]  
    Метод множителей:
    \\[ 139 = p*q + 1 = 2*69 + 1 \\]
    \\[ y = x^2 \; y^{69} = (y^3)^{23} = ((((y^2*y)^2)^2)^2)^2*((y^2*y)^2)^2*(y^2*y)^2*(y^2*y)\\\\]  
    Количество операций представлено в табл. 1
    
    
    
\\[ Таблица \; 1 - количество \; операций \; для \; бинарного \; метода \; и \; метода \; множителей \\]

|  n  | бинарный метод | Метод множителей |
|:---:|:--------------:|:----------------:|
|  60 |        8       |         9        |
| 139 |       10       |        15        |

Исходя из результатов, приведённых в таблице, можно сказать, что метод множителей выполняет либо такое же количество операций, либо гораздо большее, когда наименьший простой множитель для числа $n$ равен двойке. В противном случае алгоритм способен выдавать лучшие результаты.

2) Был реализован алгоритм Яо для вычисления приближённых аддитивных цепочек при вариировании параметра $k$. Код представлен ниже:

In [2]:
def Yao_chain(n,k):
    binar = format(n,'b') # двоичная запись числа
    lamb = len(binar) - 1 # лямбда числа n
    res = [ 2**x for x in range(0,lamb + 1)] # базовая последовательность
    base = 2**k
    new_not = notation(n,base) # число в 2**k системе счисления
    rev_not = new_not[::-1]
    elem = 0
    fin_coeff = []
    while(new_not != []):
        d = 0
        while(elem in new_not):
            new_not.remove(elem)
        for i in range(len(rev_not)):
            if rev_not[i] == elem:
                    d += base**i
        if d != 0:
            iterator = 1
            while(iterator <= elem):
                    tmp = d*iterator # значение функции d(z)
                    if tmp not in res:
                        res.append(tmp)
                    iterator *= 2 # перепрыгиваем ненужные значения
            if d*elem != 0:
                if d*elem not in res:
                    res.append(d*elem)
                fin_coeff.append(d*elem)
        if new_not != []:
               elem = min(new_not)
    if sum(fin_coeff) not in res:
        res.append(sum(fin_coeff))
        res.sort()
    for i in range(len(res)):
            res[i] = str(res[i])
    return " ".join(res)
def notation(n,base): # перевод в систему счисления с основанием base
    res = []
    cur = n + 1
    while(n >= base):
        res.append(n % base)
        n = floor(n/base)
    res.append(n)
    res = res[::-1]
    return res

n = int(input())
k = int(input())
Yao_chain(n,k)


1007
3


'1 2 4 8 16 32 40 64 65 128 130 256 260 455 512 1007'

Результаты вычислений аддитивных цепочек при различных $n$ и параметре $k$ методом Яо представлены в табл. 2

\\[ Таблица \; 2 - \; Таблица \; вычислений \; аддитивных \; цепочек \; методом \; Яо \\]

|   n  | k |                          a                          | len | min_len |
|:----:|:-:|:---------------------------------------------------:|-----|---------|
| 1007 | 3 | 1 2 4 8 16 32 40 64 65 128 130 256 260 455 512 1007 |  16 |    11   |
| 1007 | 5 |       1 2 4 8 15 16 32 64 128 256 512 992 1007      |  13 |    11   |
| 1007 | 9 |        1 2 4 8 16 32 64 128 256 495 512 1007        |  12 |    11   |
| 1024 | 3 |          1 2 4 8 16 32 64 128 256 512 1024          |  11 |    11   |
| 1024 | 5 |          1 2 4 8 16 32 64 128 256 512 1024          |  11 |    11   |
| 1024 | 9 |          1 2 4 8 16 32 64 128 256 512 1024          |  11 |    11   |
| 2001 | 3 | 1 2 4 8 16 32 64 128 256 448 512 1024 1536 2001     |  14 |    12   |
| 2001 | 5 | 1 2 4 8 16 17 32 64 128 256 512 960 1024 2001       |  14 |    12   |
| 2001 | 9 | 1 2 4 8 16 32 64 128 256 465 512 1024 1536 2001     |  14 |    12   |

Исходя из результатов таблицы, можно сделать вывод, что с увеличением параметра $k$ длина аддитивной цепочки уменьшается. Если число $n$ является степенью двойки, то длина цепочки не будет изменятся от $k$ 

3) Был реализован алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки натурального числа $n$. Код представлен ниже:

In [ ]:
import time
from sage.functions.log import logb
def vector_crashing(n):
    binar = format(n,'b') # двоичная запись n
    lamb = len(binar) - 1 # lambda(n)
    nu = binar.count('1')
    lower = lamb + floor(logb(nu,2))
    upper = lamb + nu - 1
    res = []
    m = lower
    start = time.time()
    while(res == []):
        r = [x for x in range(1,m + 1)]
        first_size = int(len(r) / 2)
        p = r[first_size::]   # измен. часть 
        r = r[:first_size:]   # неизмен. часть 
        vec = r+p
        tmp = build_chain(vec)
        a_min = tmp[0]
        a_max = tmp[len(tmp)- 1]
        if tmp[len(tmp) - 1] == n:
            res = tmp
            print("--- %s seconds ---" % (time.time() - start))
            return res
        pos_p = len(p) - 1
        pos_r = len(r) - 1
        prev_values_p = p[::]
        prev_values_r = r[::]
        flag = 0
        if a_min <= n and n <= a_max:
            while(vec.count(1) != len(vec)):  #r.count(1) != len(r)
                while(p.count(1) != len(p)): # пока можно уменьшить вектор p.count(1) != len(p)
                    p[pos_p] -= 1
                    while p[pos_p] == 0:
                        p[pos_p] = prev_values_p[pos_p]
                        pos_p -= 1
                        if pos_p == -1:
                            break
                        p[pos_p] -= 1
                    pos_p = len(p) - 1
                    if p == prev_values_p: # [1,1,...,1]
                        break
    
                    vec = r+p
                    tmp = build_chain(vec)
                    a_max = tmp[len(tmp) - 1]
                    if n == tmp[len(tmp) - 1]:
                        res = tmp
                        print("--- %s seconds ---" % (time.time() - start))
                        return res
                r[pos_r] -= 1
                while r[pos_r] == 0:
                        r[pos_r] = prev_values_r[pos_r]
                        pos_r -= 1
                        if pos_r == -1:
                            break
                        r[pos_r] -= 1
                pos_r = len(r) - 1
                pos_p = len(p) - 1
                if n < 1000:
                    p = prev_values_p[::] #
                    vec = r + p #
                if r == prev_values_r: # [1,1,...,1]
                        break
        m += 1
    print("--- %s seconds ---" % (time.time() - start))
    return res
def build_chain(vec):
    tmp = [1]
    for i in range(1,len(vec)+1):
            tmp.append(tmp[i-1] + tmp[vec[i-1] - 1])
    return tmp
n = int(input())
vector_crashing(n)

1007


Результаты вычисления минимальной звёздной цепочки алгоритмом дробления вектора индексов представлены в табл. 3

\\[ Таблица \; 3 - \; Таблица \; вычислений \; минимальных \; звёздных \; цепочек  \\]

|   n  |                            a                            | len |        time        |
|:----:|:-------------------------------------------------------:|-----|:------------------:|
| 1007 |       1 2 4 8 16 32 64 65 130 260 325 650 975 1007      |  14 |   125.861 seconds  |
| 2001 | 1 2 4 8 16 32 64 128 256 384 400 800 1600 2000 2001     |  15 |   154.212 seconds  |
| 2048 | 1 2 4 8 16 32 64 128 256 512 1024 2048                  |  12 | 0.00007415 seconds |
| 3000 | 1 2 4 8 16 32 64 128 192 200 400 800 1000 2000 3000     |  15 |   203.378 seconds  |
| 4005 | 1 2 4 8 16 32 64 128 256 512 768 800 801 1602 3204 4005 |  16 |   293.742 seconds  |

Исходя из результатов таблицы можно сделать вывод о том, что для чисел $n$ кратных степеням двойки алгоритм находит цепочку достаточно быстро, в то время как поиск звёздных цепочек для остальных чисел занимает довольно длительное время. Это связано с тем, что алгоритм основан на переборе всех возможных вариантов вектора, по которому строится цепочка.


## Выводы
В ходе выполнения практической работы были реализованы алгоритмы для создания аддитивных цепочек алгоритмом Яо и алгоритмом дробления вектора индексов. Исследования показали, что алгоритм Яо строит не самую минимальную цепочку для заданного числа $n$, а лишь какуе-то. Алгоритм дробления вектора индексов же строит минимальную звёздную цепочку, но затрачивает на это достаточно большое количество времени ввиду того, что является переборным.